components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -54,124 +67,527 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
   }
   [self setNeedsLayout];
 }
 
-- (void)setNumberOfButtons:(NSUInteger)numberOfButtons {
-  NSAssert(numberOfButtons <= kNumberOfButtonsLimit,
-           @"%@ class doesn't support more than %lu buttons", NSStringFromClass([self class]),
-           (unsigned long)kNumberOfButtonsLimit);
-  if (numberOfButtons != self.buttons.count) {
-    NSMutableArray *mutableButtons = [self.buttons mutableCopy];
-    if (numberOfButtons > self.buttons.count) {
-      for (NSUInteger index = self.buttons.count; index < numberOfButtons; ++index) {
-        MDCButton *button = [[MDCButton alloc] init];
-        [mutableButtons addObject:button];
-        [self addSubview:button];
-      }
-    } else if (numberOfButtons < self.buttons.count) {
-      for (NSUInteger index = numberOfButtons; index < self.buttons.count; ++index) {
-        MDCButton *lastObject = [mutableButtons lastObject];
-        [lastObject removeFromSuperview];
-        [mutableButtons removeLastObject];
-      }
-    }
-    self.buttons = [mutableButtons copy];
+- (UIImage *)icon {
+  return self.iconImageView.image;
+}
+
+- (void)setButtons:(NSMutableArray<MDCButton *> *_Nonnull)buttons {
+  NSAssert(buttons.count <= kNumberOfButtonsLimit, @"%@ doesn't support more than %lu buttons",
+           NSStringFromClass([self class]), (unsigned long)kNumberOfButtonsLimit);
+  _buttons = buttons;
+  for (MDCButton *button in buttons) {
+    button.translatesAutoresizingMaskIntoConstraints = NO;
+    [self.containerView addSubview:button];
   }
+  self.isLayoutDirty = YES;
+}
+
+- (void)setBackgroundColor:(UIColor *)backgroundColor {
+  [super setBackgroundColor:backgroundColor];
+  self.containerView.backgroundColor = backgroundColor;
+}
+
+- (void)setTextFont:(UIFont *)textFont {
+  self.textLabel.font = textFont;
+}
+
+- (UIFont *)textFont {
+  return self.textLabel.font;
+}
+
+- (void)setTextColor:(UIColor *)textColor {
+  self.textLabel.textColor = textColor;
+}
+
+- (UIColor *)textColor {
+  return self.textLabel.textColor;
 }
 
-- (NSUInteger)numberOfButtons {
-  return self.buttons.count;
+- (void)setIconTintColor:(UIColor *)iconTintColor {
+  self.iconImageView.tintColor = iconTintColor;
+}
+
+- (UIColor *)iconTintColor {
+  return self.iconImageView.tintColor;
 }
 
 #pragma mark - UIView overrides
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size
-                                                     textLabel:self.textLabel
-                                                imageContainer:self.iconImageViewContainer
-                                                       buttons:self.buttons];
-  return [self.layout frameSize];
+  return CGSizeMake([UIScreen mainScreen].bounds.size.width, self.layout.size.height);
 }
 
-- (void)layoutSubviews {
-  [super layoutSubviews];
+- (CGSize)intrinsicContentSize {
+  return CGSizeMake([UIScreen mainScreen].bounds.size.width, self.layout.size.height);
```

Comment:
Question: Do we know it will always be the size of the screen? Can't a banner be the size of the window, i.e. in split view mode on iPad?

Reply:
Ah. Good catch. Should I use `self.view.window.screen.bounds` here or not? Any suggestion?

Reply:
I rarely see it been used in the codebase, so that was where my confusion came from.

Reply:
Please be sure not to use screen sizes since on multi-tasking iPads the screen size is irrelevant.

Reply:
I think maybe you want to look at:
```
if (self.superview) {
    return CGSizeMake(CGRectGetWidth(self.superview.frame), self.layout.size.height);
}
return CGSizeMake(UIViewNoIntrinsicMetric, UIViewNoIntrinsicMetric)
```

Reply:
IntrinsicContentSize should never depend on superview or its position within a view hierarchy.  Please do not inspect `self.superview` to change the behavior of the method.  If the view can be any arbitrary size, then `CGSizeMake(UIViewNoIntrinsicMetric, UIViewNoIntrinsicMetric)` seems correct.  However, I suspect banners have some intrinsic height (determined by labels/buttons) and potentially a minimum intrinsic width.

Reply:
For the width, I do have a minimum width, but should I simply use `CGSizeMake(minimumWidth, aCalculatedHeight)` and expect client to set constraint to make it fill the screen width in this case? I do want to make the banner to be full screen width in both `intrinsicContentSize` and `sizeThatFits` as its real size.

Reply:
Neither `intrinsicContentSize` nor `sizeThatFits` will determine a view's size on their own.  These methods are provided explicitly to help superviews perform layout.  Both can be ignored by the superview simply setting the size of the Banner, either through constraints or `bounds`+`center` changes (assigning the frame is generally discouraged in UIKit).

Typically, `intrinsicContentSize` is used along with layout constraints to define the positioning of subviews.  If you don't provide an explicit width (or leading + trailing), then the `intrinisicContentSize` width is used. If that returns `UIViewNoIntrinsicMetric`, then you will get a constraints warning.

Similarly, `sizeThatFits:` is used when the superview is performing layout calculations (or its own `sizeThatFits:`. What is returned should be the size the view wants to be given the input size. In some cases, a view ignores the input size (like `UISwitch` and `MDCBottomNavigationBar`). In other cases, it will return the provided size. In most cases, it's a combination of both - using the provided size as an input to guide the resulting fit size.